File: spring-security-login-logger/src/main/java/com/andrewthom/springsecurityloginlogger/listener/AuthenticationSuccessfulListener.java
package com.andrewthom.springsecurityloginlogger.listener;

import org.springframework.context.ApplicationListener;
import org.springframework.security.authentication.event.AuthenticationSuccessEvent;

import com.andrewthom.springsecurityloginlogger.service.LoginAttemptService;
import org.springframework.security.web.authentication.WebAuthenticationDetails;
import org.springframework.web.context.request.RequestContextHolder;
import org.springframework.web.context.request.ServletRequestAttributes;

import javax.servlet.http.HttpServletRequest;

import static com.andrewthom.springsecurityloginlogger.listener.AuthenticationResult.SUCCESS;

public class AuthenticationSuccessfulListener implements ApplicationListener<AuthenticationSuccessEvent> {
	private LoginAttemptService service;

	private boolean log;

	public AuthenticationSuccessfulListener(LoginAttemptService service, boolean log) {
		this.service = service;
		this.log = log;
	}

	@Override
	public void onApplicationEvent(final AuthenticationSuccessEvent event) {
		if (!log) {
			return;
		}
		final String userName = event.getAuthentication().getName();
		String ipAddress = "unknown";
		HttpServletRequest currentRequest = ((ServletRequestAttributes) RequestContextHolder.currentRequestAttributes()).getRequest();
		if (currentRequest != null) {
			String ipHeader = currentRequest.getHeader("X-Forwarded-For");
			if (ipHeader != null && !ipHeader.isEmpty()) {
				ipAddress = ipHeader;
			}
		}
		if ("unknown".equals(ipAddress)) {
			try {
				ipAddress = ((WebAuthenticationDetails) event.getAuthentication().getDetails()).getRemoteAddress();
			} catch (Exception e) {
				// continue without getting IP address
			}
		}
		service.putAttemptInDatabase(userName, SUCCESS, ipAddress);
	}
}

File: spring-security-login-logger/src/main/kotlin/com/andrewthom/springsecurityloginlogger/common/listener/AuthenticationSuccessfulListener.kt
package com.andrewthom.springsecurityloginlogger.common.listener

import com.andrewthom.springsecurityloginlogger.common.listener.AuthenticationResult.SUCCESS
import com.andrewthom.springsecurityloginlogger.common.service.LoginAttemptService
import org.springframework.context.ApplicationListener
import org.springframework.security.authentication.event.AuthenticationSuccessEvent
import org.springframework.security.web.authentication.WebAuthenticationDetails
import org.springframework.web.context.request.RequestContextHolder
import org.springframework.web.context.request.ServletRequestAttributes

class AuthenticationSuccessfulListener(private val service: LoginAttemptService, private val log: Boolean) : ApplicationListener<AuthenticationSuccessEvent> {
	override fun onApplicationEvent(event: AuthenticationSuccessEvent) {
		if (!log) {
			return
		}
		val userName = event.authentication.name
		var ipAddress = "unknown"
		val currentRequest = (RequestContextHolder.currentRequestAttributes() as? ServletRequestAttributes)?.request
		if (currentRequest != null) {
			val ipHeader = currentRequest.getHeader("X-Forwarded-For")
			if (!ipHeader.isNullOrEmpty()) {
				ipAddress = ipHeader
			}
		}
		if ("unknown" == ipAddress) {
			ipAddress = (event.authentication.details as? WebAuthenticationDetails)?.remoteAddress ?: "unknown"
		}
		service.logAttempt(userName, SUCCESS, ipAddress)
	}
}
